### ISSUE TITLE
Device selection on export on multi-gpu systems

### ISSUE THREAD

**Issue by liwtw** [ultralytics/ultralytics#19310]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Greatings! 🚀
Sorry for my English. I ran into issue (latest version, February, 19th) when choosing GPU to export on NVIDIA mGPU setup:

```
DEVICE0 = "cuda:1"
torch.set_default_device(device=DEVICE0)
with torch.cuda.device(device=DEVICE0):
    model = YOLO("yolo11m.pt")
    model.export(format="engine", half=True, imgsz=TRACK_HW, batch=BATCH_SIZE, dynamic=True, device=DEVICE0)
```

I selected second (:1) gpu, but got a usage on first (:0) one. nvidia-smi showed a full load on first gpu with a small one on second.

utils/torch_utils.py

```
    if not cpu and not mps and torch.cuda.is_available():  # prefer GPU if available
        devices = device.split(",") if device else "0"  # i.e. "0,1" -> ["0", "1"]
        n = len(devices)  # device count
        if n > 1:  # multi-GPU
            if batch < 1:
                raise ValueError(
                    "AutoBatch with batch<1 not supported for Multi-GPU training, "
                    "please specify a valid batch size, i.e. batch=16."
                )
            if batch >= 0 and batch % n != 0:  # check batch_size is divisible by device_count
                raise ValueError(
                    f"'batch={batch}' must be a multiple of GPU count {n}. Try 'batch={batch // n * n}' or "
                    f"'batch={batch // n * n + n}', the nearest batch sizes evenly divisible by {n}."
                )
        space = " " * (len(s) + 1)
        for i, d in enumerate(devices):
            s += f"{'' if i == 0 else space}CUDA:{d} ({get_gpu_info(i)})\n"  # bytes to MB
        arg = "cuda:0"
```

The line leads to bug: ```arg = "cuda:0"```

I suppose it should be: ```arg = f"cuda:{device}"```

### Environment

Package                   Version
------------------------- ------------
addict                    2.4.0
aiohappyeyeballs          2.4.4
aiohttp                   3.11.11
aiosignal                 1.3.2
albucore                  0.0.23
albumentations            2.0.2
annotated-types           0.7.0
anyio                     4.8.0
attrs                     25.1.0
bcrypt                    4.2.1
certifi                   2025.1.31
cffi                      1.17.1
chardet                   5.2.0
charset-normalizer        3.4.1
click                     8.1.8
coloredlogs               15.0.1
contourpy                 1.3.1
cryptography              44.0.0
cycler                    0.12.1
fastapi                   0.115.6
filelock                  3.17.0
flatbuffers               25.1.24
fonttools                 4.55.7
frozenlist                1.5.0
fsspec                    2025.2.0
geographiclib             2.0
greenlet                  3.1.1
h11                       0.14.0
huggingface-hub           0.27.1
humanfriendly             10.0
idna                      3.10
Jinja2                    3.1.5
jsonschema                4.23.0
jsonschema-specifications 2024.10.1
jwt                       1.3.1
kiwisolver                1.4.8
lap                       0.5.12
lightning-utilities       0.11.9
MarkupSafe                3.0.2
matplotlib                3.10.0
mpmath                    1.3.0
msgpack                   1.1.0
multidict                 6.1.0
networkx                  3.4.2
numpy                     2.1.1
nvidia-cublas-cu12        12.6.4.1
nvidia-cuda-cupti-cu12    12.6.80
nvidia-cuda-nvrtc-cu12    12.6.77
nvidia-cuda-runtime-cu12  12.6.77
nvidia-cudnn-cu12         9.5.1.17
nvidia-cufft-cu12         11.3.0.4
nvidia-curand-cu12        10.3.7.77
nvidia-cusolver-cu12      11.7.1.2
nvidia-cusparse-cu12      12.5.4.2
nvidia-cusparselt-cu12    0.6.3
nvidia-nccl-cu12          2.21.5
nvidia-nvjitlink-cu12     12.6.85
nvidia-nvtx-cu12          12.6.77
onnx                      1.17.0
onnxruntime-gpu           1.20.1
onnxslim                  0.1.48
opencv-python             4.11.0.86
opencv-python-headless    4.11.0.86
openvino                  2025.0.0
openvino-telemetry        2025.0.0
packaging                 24.2
pandas                    2.2.3
pillow                    11.1.0
pip                       24.3.1
propcache                 0.2.1
protobuf                  5.29.3
psutil                    6.1.1
psycopg2-binary           2.9.10
py-cpuinfo                9.0.0
pyarrow                   19.0.0
pycparser                 2.22
pydantic                  2.10.5
pydantic_core             2.27.2
PyJWT                     2.10.1
pyparsing                 3.2.1
pysrt                     1.1.2
python-dateutil           2.9.0.post0
python-dotenv             1.0.1
python-magic              0.4.27
python-multipart          0.0.20
pytorch-lightning         2.5.0.post0
pytz                      2025.1
PyYAML                    6.0.2
pyzmq                     26.2.0
ray                       2.40.0
referencing               0.36.2
requests                  2.32.3
rpds-py                   0.22.3
safetensors               0.5.2
scipy                     1.15.1
seaborn                   0.13.2
setuptools                75.8.0
simsimd                   6.2.1
six                       1.17.0
sniffio                   1.3.1
SQLAlchemy                2.0.37
sqlmodel                  0.0.22
starlette                 0.41.3
stringzilla               3.11.3
sympy                     1.13.1
tensorboardX              2.6.2.2
tensorrt                  10.7.0.post1
tensorrt_cu12             10.7.0.post1
tensorrt-cu12-bindings    10.7.0.post1
tensorrt-cu12-libs        10.7.0.post1
timm                      1.0.14
torch                     2.6.0+cu126
torch_tensorrt            2.6.0+cu126
torchaudio                2.6.0+cu126
TorchCodec                0.2.0+cu126
torchmetrics              1.0.3
torchvision               0.21.0+cu126
tqdm                      4.67.1
triton                    3.2.0
typing_extensions         4.12.2
tzdata                    2025.1
ultralytics               8.3.76
ultralytics-thop          2.0.14
urllib3                   2.3.0
uvicorn                   0.34.0
websockets                14.2
wheel                     0.45.1
yarl                      1.18.3

### Minimal Reproducible Example

```
DEVICE0 = "cuda:1"
torch.set_default_device(device=DEVICE0)
with torch.cuda.device(device=DEVICE0):
    model = YOLO("yolo11m.pt")
    model.export(format="engine", half=True, imgsz=TRACK_HW, batch=BATCH_SIZE, dynamic=True, device=DEVICE0)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liwtw, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many of the most common questions may already be addressed there.

If this is a 🐛 Bug Report, thank you for the detailed description! To help us investigate further, please ensure that you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). While you have already included relevant code snippets and environment details, having an isolated, simple script that consistently reproduces the issue would be extremely helpful for debugging.

If this is an ❓ Implementation or Usage Question, please provide as much context as possible, including dataset details (if applicable), training logs, and how you are executing your scripts. Kindly ensure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the growing Ultralytics community where it suits you best:
- Real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧  
- Long-form discussions on [Discourse](https://community.ultralytics.com/)  
- Engage with the Ultralytics subreddit on [Reddit](https://reddit.com/r/Ultralytics)  

## Upgrade

Please ensure you are using the latest `ultralytics` package. Update it in a clean [**Python>=3.8**](https://www.python.org/) environment with the required [dependencies](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) to confirm the issue still persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following verified environments, ensuring pre-installed [Python](https://www.python.org/), [PyTorch](https://pytorch.org/), [CUDA](https://developer.nvidia.com/cuda-zone), and [CUDNN](https://developer.nvidia.com/cudnn):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM via the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI as per the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image** as shown in the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests run every 24 hours to verify proper operation across [modes](https://docs.ultralytics.com/modes/) and [tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This is an automated response to help us get started quickly. An Ultralytics engineer will review your issue and respond promptly. Thank you for bringing this to our attention! 😊

**Comment by Y-T-G** [comment]
You just need to use the `model.export()` line. Remove the other lines.

`CUDA_VISIBLE_DEVICES` is set to the device you pass.:

https://github.com/ultralytics/ultralytics/blob/fdeb04e7bc8be5cef53c85cbfc7464b5796cb067/ultralytics/utils/torch_utils.py#L179

After setting it, 0 refers to the device you passed since that's the first GPU now, so it's correct.

**Comment by liwtw** [comment]
It works! Thank you!

**Comment by glenn-jocher** [comment]
@liwtw great! Glad to hear the issue is resolved. Let us know if you have any other questions.
